src/main.rs

```diff
@@ -36,7 +36,14 @@ pub mod expansion;
 pub struct Command {
     pub name: &'static str,
     pub help: &'static str,
-    pub main: Box<Fn(&Vec<String>, &mut BTreeMap<String, String>, &mut Vec<Mode>)>,
+    pub main: Box<Fn(&Vec<String>, &mut Shell)>,
```

Comment:
I think I do want to do this because otherwise the shell is forever borrowed by the function that creates the closures. The closures last for pretty much the whole time the main process is running, so having the shell stuck there won't work.